teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-<tekton-generated-suffix>
+spec:
+  ref:
+    apiVersion: example.dev/v0
+    kind: Example
+    name: my-example
+  params:
+  - name: task-param
+    value: hello
+```
+
+The Custom Task author's controller will be watching for creations of `Run`s of this type, and can invoke a Tekton-provided Go package to resolve parameters in string-typed fields, using reflection. This will give them a struct value like:
+
+```
+{
+  "spec": {
+    "params": [ … ],
+    "task-param": "hello, old friend"
+  }
+}
+```
+
+### Result Reporting
+
+Custom Task controllers can report any results via the `Run`'s `.status.results` field:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-blah-blah
+spec:
+  … 
+status:
+  results:
+  - name: first-name
+    value: Bob
+```
+
+If their custom defined CRD type has a `.results` field of a compatible type, `Pipeline` validation can take that into account when validating inputs and outputs between Tasks (custom or traditional) in a `Pipeline`.
+
+Controllers can report any results, regardless of whether the underlying CRD object declared them.
+
+### Cancellation
+
+To support cancellation of `Run`s, when a `PipelineRun` is cancelled, the `PipelineRun`  controller will attempt to update any ongoing `Run`s' `.spec.status` to `"Cancelled"`, and update `.status.conditions` to signal unsuccesful execution.
+
+A Custom Task author can watch for this status update and take any corresponding actions (e.g., cancel a cloud build, stop the waiting timer, tear down the approval listener).
+
+Supporting cancellation is optional but recommended.
+
+### Timeout
+
+Today, users can specify a timeout for a component `Task` of a `Pipeline` (see [`PipelineTask.Timeout`](https://godoc.org/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#PipelineTask)). The `Run` type will specify a `Timeout` field to hold this value when created as part of a `PipelineRun` (or when `Run`s are created directly). Custom Task authors can read and propagate this field if desired, but a Tekton-owned controller will be responsible for tracking this timeout and updating timed out `Run`s to signal unsuccessful execution.
+
+A Custom Task author can watch for this status update and take any corresponding actions (e.g., cancel a cloud build, stop the waiting timer, tear down the approval listener).
+
+Supporting timeouts is optional but recommended.
+
+### CLI and Dashboard Support
+
+At the very least, the `tkn` CLI and Tekton Dashboard should have some way to display basic information about Custom Tasks, even if it's just a dump of the YAML. Solving a complete holistic plugin model for Go binaries and web front-ends expands the scope of this work too broadly, but at least providing Custom Task authors some basic support in Tekton's native tooling is better than nothing.
```

Comment:
What about a `run` subcommand for listing runs?

Reply:
BTW the bit about surfacing the children `runs` was thinking forward to `tkn run logs` for Pipeline/Task/Matrix/Custom/...

Reply:
Blegh, I kind of just don't want to think about CLI support at all, let alone plumbing through `tkn logs` support, which would insinuate some way to tell the CLI that you do end up running Pods. Blegh.

Reply:
This is why I tagged @n3wscott elsewhere, since he's trying to motivate custom subresources in K8s and `/logs` is one that I believe has come up.